src/dotnet/commands/dotnet-compile/ManagedCompiler.cs

```diff
@@ -181,31 +184,15 @@ public override bool Compile(ProjectContext context, CompilerCommandApp args)
 
             _scriptRunner.RunScripts(context, ScriptNames.PreCompile, contextVariables);
 
-            var result = _commandFactory.Create($"compile-{compilerName}", new[] { "@" + $"{rsp}" })
-                .OnErrorLine(line =>
-                {
-                    var diagnostic = ParseDiagnostic(context.ProjectDirectory, line);
-                    if (diagnostic != null)
-                    {
-                        diagnostics.Add(diagnostic);
-                    }
-                    else
-                    {
-                        Reporter.Error.WriteLine(line);
-                    }
-                })
-                .OnOutputLine(line =>
-                {
-                    var diagnostic = ParseDiagnostic(context.ProjectDirectory, line);
-                    if (diagnostic != null)
-                    {
-                        diagnostics.Add(diagnostic);
-                    }
-                    else
-                    {
-                        Reporter.Output.WriteLine(line);
-                    }
-                }).Execute();
+            // Cache the reporters before invoking the command in case it is a built-in command, which replaces
```

Comment:
Unfortunately it can't.

The reason this place is special is because we are calling Reporter.Write during OnErrorLine and OnOutputLine handlers. At this point, Reporter.Out is going to write to the interception TextWriter, which will then raise OnOutputLine again, and you'll get into an infinite recursion.

BuiltInCommand writes to console -> OnOutputLine handler is invoked -> Writes to Reporter.Out -> OnOutputLine handler invoked -> ...

Since this callsite expicitly is handling the BuiltInCommand's output and wants to re-write that output to the console, it needs to cache the Reporter that is pointing to the real stdout